cilium: Rework neighbor handling

borkmann opened the pull request:
See commits.

borkmann committed:
neigh: Rework ARP handling to let the kernel do the resolution

Let the kernel do the work instead of open coding ARP resolution in the agent.
Use NTF_EXT_MANAGED iff possible from kernel side. This work was presented also
during this year's Plumbers conference:

  https://linuxplumbersconf.org/event/11/contributions/953/

And recently it has been merged into the kernel / iproute2:

  https://lore.kernel.org/netdev/20211011121238.25542-1-daniel@iogearbox.net/
  https://lore.kernel.org/netdev/20211013132140.11143-1-daniel@iogearbox.net/
  https://lore.kernel.org/netdev/20211025154728.2161-1-daniel@iogearbox.net/

NTF_EXT_MANAGED flagged entries are L3 addresses of next hops which we push
into the kernel without a L2 lladdress. These are being placed onto a managed
neighbor list which the kernel periodically walks and auto-refreshes such that
the entry is kept in REACHABLE state if possible. So aside from agent pushing
these down (and deleting them upon node delete events), there's no further
work required such as refresh controller, etc. Upon start, the agent probes
availability of the feature and sets option.Config.ARPPingKernelManaged when
possible to transparently opt into it.

For older kernels, we also reuse kernel infrastructure instead of open coding
an arping library by probing neighbor entries via NTF_USE flag. Similarly,
these entries we push down are L3 addresses of next hops without L2 lladdress.
The NTF_USE flag causes the kernel to trigger a one-time refresh of the entry
if it was in a STALE state. In contrast to NTF_EXT_MANAGED we cannot be fully
hands off in this case, since this still requires the agent's neighbor refresh
controller to trigger re-resolution. NTF_EXT_MANAGED avoids the extra netlink
churn when having lots of backends.

In either case, we push entries down via NTF_EXT_LEARNED flag such that these
are not GC'ed by the kernel. This is important since such neighbor entries are
used in fast-path, for example, to forward LB traffic to backends via XDP or
tc BPF programs, and at that layer it is not possible to resolve these on demand
through a slow-path. If kernel neighbor GC would evict them under pressure, then
next requests to such backends would fail. Thus, this preserves part of the old
NUD_PERMANENT properties, but for dynamically resolved entries.

Note that in contrast to NUD_PERMANENT entries, any active node traffic to such
neighbor from the networking stack will also update/refresh a potential STALE to
REACHABLE state. Also, removing the ARP library through this infrastructure in
here also has the advantage that we can now support IPv6 neighbor resolution with
the same mechanism. The agent is currently lacking this with the ARP library it
uses, so backends may not be able to get resolved properly. The IPv6 support
will be added separately and requires some more code refactoring.

Some more agent/implementation specific notes in terms of the ARP rework:

We still keep the next hop tracking in the agent in place since these need to be
reference counted there in case, for example, backend IP are used for multiple
services and they are direct next hops. We store neigh entries directly in the
agent, but with the transition we don't store neigh.HardwareAddr anymore. This
is okay as the way we update/delete them, they are not required, e.g. deletes
don't look at the lladdress in the kernel for a match.

For older kernels the NTF_EXT_LEARNED | NTF_USE handling is a bit cumbersome,
since the kernel has some behavioral quirks. We fixed them in above patches,
but most kernels don't have them at this point. Upon creation of new neighbor
entries, NTF_EXT_LEARNED flag is not being propagated, so we need to first
create a new entry e.g. in NUD_NONE pseudo-state with NTF_EXT_LEARNED and then
migrate this off from NUD_NONE with another request combined with NTF_USE to
trigger resolution. For already existing dynamic entries, this is not needed.

The same situation is when we migrate NUD_PERMANENT neighbor states over to
dynamic ones in NodeCleanNeighbors(). Older kernels don't have a direct path
to switch to NTF_EXT_LEARNED | NTF_USE, so we need to graduate them through
NUD_REACHABLE with NTF_EXT_LEARNED and then triggering another request with
NTF_USE flag. However, that migration is just a one-time event.

Fixes: #14962
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
neigh: Remove ARPing library from the agent

Not needed anymore, so lets retire this hack. It was/is problematic in
multiple ways as outlined in our Plumbers talk, and still not in perfect
shape:

  https://linuxplumbersconf.org/event/11/contributions/953/

See also GH issue #14962 with additional details. Moreover IPv6 support
was missing and in order to resolve it, lets fully rely on what the kernel
does best anyway.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
neigh: Update unit tests to probe for REACHABLE state

And also remove the neighByNextHop[] hardware address test since the
latter will be zero in there.

We reduce base reachable time from 30s down to 2.5s in order to avoid
for the ARP test taking too long. Also, given we bump the timeout
for the neigbor entry check from 5 to 20s, we're also bumping the
GOTEST_BASE timeout from 360s to 420s just to be on the safe side.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
neigh: Deprecate bpf-lb-bypass-fib-lookup and always rely on fib lookup

This commit removes nodeport_lb_hairpin() optimization and instead always
relies on the kernel's neighbor table. Given managed neighbor entries are
entriely kernel-managed there is no mechanism where the BPF neighbor map
would get an event upon L2 address change to flush the old/stale entry.

Similarly, in rev_nodeport_lb{4,6}() we let the fib_lookup() take precedence
over the NODEPORT_NEIGH{4,6} BPF map consultation. The latter now soley
records original L2 (client) addresses from inbound requests.

Meaning, when for the NodePort reverse path the fib_lookup() fails e.g.
due to a client in the same L2 that the kernel's neighbor table has never
seen before (but was previously processed by XDP/tc BPF layer and forwarded
from there to a backend), then such entry will be present in the BPF
NODEPORT_NEIGH{4,6} map.

Overall this change makes the traffic handling more robust, and removes the
caching behavior where entries can become stale in the context of purely
(kernel-)managed neighbors (be it via NTF_USE or NTF_EXT_MANAGED).

This also fixes issues like #17693 where the assumption that the destination
MAC of the outbound frame is always set to the previously "learnt" source MAC
of the inbound frame is not always true. Given the inversed precedence, this
implicitly addresses this case as well.

Fixes: #17693
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

ldelossa commented on go.mod:
```diff
@@ -225,6 +225,9 @@ replace (
 	github.com/optiopay/kafka => github.com/cilium/kafka v0.0.0-20180809090225-01ce283b732b
 
+	// Using cilium/netlink until neighboring patches merged upstream
+	github.com/vishvananda/netlink => github.com/cilium/netlink v1.0.1-0.20211029134139-e692b68616ee
```
+1 to this. there are other issues in the netlink library which I'm fishing out. We may want to maintain this fork as a permanent replacement, since upstream seems a bit unmaintained (last merges were in September and valid issues date back to 2018 without fixes)

borkmann replied:
Agree, still having hopes it will get merged soonish, but it's annoying that it takes usually long.

ldelossa commented on daemon/cmd/daemon.go:
```diff
@@ -316,6 +316,8 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 		return nil, nil, fmt.Errorf("unable to configure API rate limiting: %w", err)
 	}
 
+	probeManagedNeighborSupport()
```
comments here?

ldelossa commented on daemon/cmd/daemon_main.go:
```diff
@@ -229,7 +229,7 @@ func initializeFlags() {
 	option.BindEnv(option.AnnotateK8sNode)
 
-	flags.Duration(option.ARPPingRefreshPeriod, 5*time.Minute, "Period for remote node ARP entry refresh (set 0 to disable)")
+	flags.Duration(option.ARPPingRefreshPeriod, 40*time.Second, "Period for remote node ARP entry refresh (set 0 to disable)")
```
why such a drop? because kernel will perform refresh with more efficiency ?

borkmann replied:
I've slightly changed it to the kernel's default, added pkg/defaults entry w/ comment.

ldelossa commented on daemon/cmd/daemon_main.go:
```diff
@@ -1766,14 +1766,20 @@ func runDaemon() {
+		// If we came from an agent upgrade, migrate entries.
+		d.datapath.Node().NodeCleanNeighbors(true)
+		// Start periodical refresh of the neighbor table from the agent if needed.
+		if option.Config.ARPPingRefreshPeriod != 0 && option.Config.ARPPingKernelManaged {
```
this does not allow for simply setting "ARPPingKernelManaged" and no "ARPPingRefreshPeriod", however we just blast away "ARPPingRefreshPeriod"

Would it be more sane just to look for "ARPPingKernelManaged" alone  ?

jrfastab replied:
There is likely no reason to check RefreshPeriod != 0 to zero. Setting 0 to 0 doesn't change anything.

ldelossa commented on pkg/datapath/linux/node.go:
```diff
@@ -679,6 +671,7 @@ func getSrcAndNextHopIPv4(nodeIPv4 net.IP) (srcIPv4, nextHopIPv4 net.IP, err err
 	var link netlink.Link
+
 	n.neighLock.Lock()
 	if n.neighDiscoveryLink == nil || reflect.ValueOf(n.neighDiscoveryLink).IsNil() {
```
is reflect required here ? Are we ever storing `(*T)(nil)` in this interface or do we keep it nil unless populated with a valid interace?

borkmann replied:
@brb might know best given he wrote it? :) Anyway, code in PR doesn't change it.

ldelossa commented on pkg/datapath/linux/node.go:
```diff
@@ -716,12 +711,14 @@ func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeType
-						logfields.HardwareAddr: neigh.HardwareAddr,
-						logfields.LinkIndex:    neigh.LinkIndex,
-					}).WithError(err).Info("Unable to remove neighbor entry")
+						logfields.NextHop:   neigh.IP,
```
Does this break the possibility of more advanced setups with arp load balancing/sharing. I.e. do we have any current use cases where two layer 2 devices share an ip and coordinate who responds to arp?

borkmann replied:
I'm not aware we support such case today with the old infra, by fully relying on kernel side, it might be easier to implement going forward.

ldelossa commented on Documentation/concepts/ebpf/maps.rst:
```diff
@@ -36,7 +36,7 @@ Egress Policy            endpoint         16k             Max 16k endpoints acro
 line options for ``cilium-agent``. A given capacity can be set using
 ``--bpf-ct-global-tcp-max``, ``--bpf-ct-global-any-max``,
-``--bpf-nat-global-max``, ``--bpf-neigh-global-max``, ``--bpf-policy-map-max``,
+``--bpf-nat-global-max``, ``--bpf-policy-map-max``,
```
>In contrast to NTF_EXT_MANAGED we cannot be fully
hands off in this case, since this still requires the agent's neighbor refresh
controller to trigger re-resolution. NTF_EXT_MANAGED avoids the extra netlink
churn when having lots of backends.

From your second commit it sounds like on older kernels the neighbor entries will not be auto-refreshed. Does this still mean its okay to disable the BPF map caches across the board ?

borkmann replied:
> Does this still mean its okay to disable the BPF map caches across the board ?

Yes, still okay to remove them. It just means, we'll always rely on fib lookup to provide the L2 addresses which is fine.

pchaigno replied:
But what's the impact for older kernels of removing the maps then? Commit c46c0ed0e7d023208c7f162090694e9c3d1c09d4, which introduced the maps, stated:

> in the NodePort case, when a never seen before client sends a request, a receiving node does not create the ARP entry for the CLIENT_IP => REQUEST_SMAC mapping. This makes the fib_lookup to fail.

Is that not an issue anymore? Is that because we don't use NUD_PERMANENT anymore as stated in the main commit :point_down:?

> Note that in contrast to NUD_PERMANENT entries, any active node traffic to such neighbor from the networking stack will also update/refresh a potential STALE to REACHABLE state.

borkmann replied:
Yeah, if there are clients in the same L2 which are not tracked on the LB / middle node's neighbor table, we do need the map indeed as otherwise replies for the snat mode will fail to populate the clients mac. Also discussed this with @brb today in the morning. I'm currently changing the neighbor map in the following way: i) we retain tracking client macs, ii) for backends the fib lookup will be used. On replies, we use the fib lookup and if it fails we can check the BPF neighbor map. So still the neighbor map function as a precache is removed. This will also implicitly solve @tklauser's remarks on the broken calculation for dynamic map sizing.

ldelossa reviewed (changes requested):
commit message has typo:

```suggestion
is okay as they way we update/delete them, they are not required, e.g. deletes
             ^ /s/they/the
```

Questions and comments about configuration.

borkmann committed:
docker, runtime: Update iproute2 and bpftool images

Update bpftool for kernel feature probing as well as iproute2 for
dumping support of managed neighbor entries.

xref: https://github.com/cilium/image-tools/pull/148
xref: https://github.com/cilium/image-tools/pull/147

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
update cilium-{runtime,builder}

This is to incorporate new iproute2 & bpftool image into cilium-runtime.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

jrfastab commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -430,6 +430,25 @@ func initKubeProxyReplacementOptions() (bool, error) {
+	//   856c02dbce4f ("bpf: Introduce helper bpf_get_branch_snapshot")
+	// This is a bit of a workaround given feature probing for netlink
+	// neighboring subsystem is cumbersome. The commit was added in the
+	// same release as managed neighbors, that is, 5.16+.
```
I like it. clever.

borkmann committed:
neigh: Consider neighbor entries for multipath routes

Use status of existing neighbor entry when determining nexthop for multipath
routes. If disabled, neighbor information is not used and packets could be
directed to a failed nexthop which is the default behavior by the kernel for
compat reasons. During the discussion around managed neighbor entries, the
question came up on whether we make use of fib_multipath_use_neigh. Enabling
this setting is definitely recommended and will help bpf_fib_lookup() helper
in our datapath.

Link: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=a6db4494d218c2e559173661ee972e048dc04fdd
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

tklauser commented on bpf/lib/nodeport.h:
```diff
@@ -58,25 +58,6 @@ bpf_skip_nodeport(struct __ctx_buff *ctx)
-} NODEPORT_NEIGH4 __section_maps_btf;
-#endif /* ENABLE_IPV4 */
-
-#ifdef ENABLE_IPV6
```
I guess it doesn't matter much as it is just a type definition, but do we still need the `#ifdef ENABLE_IPV6` here (along with the corresponding `#endif`) as we presumably only use `dsr_opt_v6` when IPv6 is enabled?

borkmann replied:
Removed it since it doesn't matter for struct definitions if not used.

tklauser reviewed (commented):
Didn't do a full review yet, mainly focussed on changes releated to dynamic BPF map sizing, agent options and vendoring so far.

Regarding the options, the following changes in `pkg/option/config.go` would also make sense to remove mentions the neighbor from comments. I cannot comment on the actual lines as they are not changed, thus the following incremental diff:

```diff
diff --git a/pkg/option/config.go b/pkg/option/config.go
index 7dbe27039dc7..f97231d0551f 100644
--- a/pkg/option/config.go
+++ b/pkg/option/config.go
@@ -536,7 +536,7 @@ const (

        // MapEntriesGlobalDynamicSizeRatioName is the name of the option to
        // set the ratio of total system memory to use for dynamic sizing of the
-       // CT, NAT, Neighbor and SockRevNAT BPF maps.
+       // CT, NAT and SockRevNAT BPF maps.
        MapEntriesGlobalDynamicSizeRatioName = "bpf-map-dynamic-size-ratio"

        // LimitTableAutoGlobalTCPMin defines the minimum TCP CT table limit for
@@ -1314,7 +1314,7 @@ type DaemonConfig struct {
        MonitorAggregationFlags uint16

        // BPFMapsDynamicSizeRatio is ratio of total system memory to use for
-       // dynamic sizing of the CT, NAT, Neighbor and SockRevNAT BPF maps.
+       // dynamic sizing of the CT, NAT and SockRevNAT BPF maps.
        BPFMapsDynamicSizeRatio float64

        // NATMapEntriesGlobal is the maximum number of NAT mappings allowed
```

The fact that the neighbor map is removed and is thus no longer considered for dynamic sizing will likely also change the default sizes for the other maps which are still included in dynamic sizing. Maybe we should mention this in the 1.11 upgrade notes as it could lead to the maps getting re-created on upgrade?

ldelossa reviewed (approved):
feedback from my review resolved. approving.

qmonnet commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -18,7 +18,7 @@ cilium-agent [flags]
       --annotate-k8s-node                                    Annotate Kubernetes node (default true)
       --api-rate-limit map                                   API rate limiting configuration (example: --rate-limit endpoint-create=rate-limit:10/m,rate-burst:2) (default map[])
-      --arping-refresh-period duration                       Period for remote node ARP entry refresh (set 0 to disable) (default 5m0s)
+      --arping-refresh-period duration                       Period for remote node ARP entry refresh (set 0 to disable) (default 30s)
```
We still set this to 5 minutes in install/kubernetes/cilium/values.yaml. Should that value be updated?

borkmann replied:
Good catch. I would rather remove anything that is explicitly set there so it sticks to agent default. Will fix, thanks!

borkmann replied:
Should be done now, pls double check. Thanks!

qmonnet reviewed (approved):
Looks good to me, nice work!
One question below.

borkmann committed:
neigh, helm: Update l2NeighDiscovery.arping-refresh-period's default

Ideally commit da746075e6f5 ("helm: Expose l2 neigh discovery related agent
flags") would never have exposed a default value of 5m to helm. Just relying
on agent default would have been enough w/o setting a default value again in
here, so that it could have been more easily changed. This commit in here
updates it to 30 seconds which is in line with the kernel's base reachable
time for arp/ndisc, and matches the agent's defaults.ARPBaseReachableTime as
well. Also, improve the vague description a bit.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
neigh, doc: Rework KPR neighbor resolution section

Document the old and new behavior of the Cilium agent in the KPR guide's
Neighbor Discovery section.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

jrfastab commented on daemon/cmd/daemon_main.go:
```diff
@@ -1766,14 +1766,17 @@ func runDaemon() {
+		d.datapath.Node().NodeCleanNeighbors(false)
+	} else {
+		// If we came from an agent upgrade, migrate entries.
+		d.datapath.Node().NodeCleanNeighbors(true)
```
stupid nit, I would probably name these NoceCleanNeighbors() for the delete and NodeMigrateNeighbors(). But its clear enough as-is imo.

jrfastab replied:
then we would also have to stub the fakes so likely not worth the rename.

jrfastab commented on daemon/cmd/daemon_main.go:
```diff
@@ -1766,14 +1766,17 @@ func runDaemon() {
+		// If we came from an agent upgrade, migrate entries.
+		d.datapath.Node().NodeCleanNeighbors(true)
+		// Start periodical refresh of the neighbor table from the agent if needed.
+		if option.Config.ARPPingRefreshPeriod != 0 && !option.Config.ARPPingKernelManaged {
```
nit... is there a reason to have the '&&' check seems ARPPingRefreshPeriod should be sufficient.

Is there a reason to have ARPPingRefreshPeriod != 0 and ARPPingKernelManaged set?

borkmann replied:
ARPPingRefreshPeriod is set to 30s by default, unless the user explicitly turns it off. ARPPingKernelManaged is auto-detected inside the agent. So if available due to the probe succeeding, then we don't need to start the refresh controller at all.

jrfastab reviewed (approved):
Amazing. ACK! Having struggled with this in the past this is a big improvment.

Couple minor nits feel free to ignore them.

pchaigno commented on pkg/datapath/linux/node.go:
```diff
@@ -1497,7 +1487,7 @@ func (n *linuxNodeHandler) NodeNeighborRefresh(ctx context.Context, nodeToRefres
 // discovery link interfaces. It should be used when the agent changes the state
 // from `n.enableNeighDiscovery = true` to `n.enableNeighDiscovery = false`.
-func (n *linuxNodeHandler) NodeCleanNeighbors() {
+func (n *linuxNodeHandler) NodeCleanNeighbors(migrateOnly bool) {
```
The function comment above needs an update.

pchaigno commented on pkg/datapath/linux/node.go:
```diff
@@ -1544,31 +1533,80 @@ func (n *linuxNodeHandler) NodeCleanNeighbors() {
+		// the kernel eventually. Older Cilium versions might have left-
+		// overs installed as NUD_PERMANENT.
+		if (neigh.State&netlink.NUD_PERMANENT) == 0 &&
+			(neigh.Flags&netlink.NTF_EXT_LEARNED) == 0 {
```
Nit: I don't think the parenthesis are required here in Go:
```suggestion
		if neigh.State&netlink.NUD_PERMANENT == 0 &&
			neigh.Flags&netlink.NTF_EXT_LEARNED == 0 {
```

pchaigno commented on pkg/datapath/linux/node.go:
```diff
@@ -1544,31 +1533,80 @@ func (n *linuxNodeHandler) NodeCleanNeighbors() {
+		}
+		if migrateOnly {
+			// We only care to migrate NUD_PERMANENT over to dynamic
+			// state entires with NTF_EXT_LEARNED.
```
Nit:
```suggestion
			// state entries with NTF_EXT_LEARNED.
```

pchaigno commented on pkg/datapath/linux/node.go:
```diff
@@ -1544,31 +1533,80 @@ func (n *linuxNodeHandler) NodeCleanNeighbors() {
+						logfields.Device:    linkName,
+						logfields.LinkIndex: l.Attrs().Index,
+						"neighbor":          fmt.Sprintf("%+v", neigh),
+					}).Info("Unable to replace new next hop")
```
Shouldn't this be logged as an error?

pchaigno commented on pkg/datapath/linux/node.go:
```diff
@@ -679,6 +671,7 @@ func getSrcAndNextHopIPv4(nodeIPv4 net.IP) (srcIPv4, nextHopIPv4 net.IP, err err
 // sends the ARP request anyway.
 func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeTypes.Node, refresh bool) {
```
The function comment needs to be updated.

pchaigno commented on pkg/datapath/linux/node.go:
```diff
@@ -639,33 +635,29 @@ func (n *linuxNodeHandler) encryptNode(newNode *nodeTypes.Node) {
+		return nil, fmt.Errorf("remote node IP is non-routable")
 	}
 
 	// Use the first available route by default
```
I think this comment can now be removed since AFAIU it only meant to explain the `srcIPv4` selection.

pchaigno commented on pkg/datapath/linux/node.go:
```diff
@@ -743,75 +740,67 @@ func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeType
+		if err := netlink.NeighSet(&neighInit); err != nil {
+			scopedLog.WithError(err).WithFields(logrus.Fields{
+				"neighbor": fmt.Sprintf("%+v", neighInit),
+			}).Info("Unable to insert new next hop")
```
I know this isn't really introduced by this pull request, but do you know why all those "Unable to insert/remove next hop" messages are logged as Info? It seems they should at least be Warnings.

borkmann replied:
(Cc @brb) It was reduced due to below commit:
```
commit 2de65d0e77df7d958ace1839c595261877e4988d
Author: Martynas Pumputis <m@lambda.lt>
Date:   Tue Mar 9 10:37:33 2021 +0100

    node-neigh: Reduce arping related log msg's level
    
    It is not the end of the world if any arping related operation fails
    (e.g. frequent connections between nodes ensure the presence of
    relevant L2 entries in the neigh table). So, decrease the log level of
    the log msgs and change the log subsystem to "node-neigh-debug".
    
    Signed-off-by: Martynas Pumputis <m@lambda.lt>
```
The main driver was probably this hunk:
```
@@ -734,7 +733,7 @@ func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeType
        if nextHopIsNew || refresh {
                hwAddr, err := arp.PingOverLink(n.neighDiscoveryLink, srcIPv4, nextHopIPv4)
                if err != nil {
-                       scopedLog.WithError(err).Error("arping failed")
+                       scopedLog.WithError(err).Info("arping failed")
                        metrics.ArpingRequestsTotal.WithLabelValues(failed).Inc()
                        return
                }
```
I think it's okay to try and have the Netlink-related operations as Warn or Errors. NeighDel (from deleteNeighbor) should be on info however.

joestringer replied:
I saw this go past my email inbox and thought it's a good time to plug the WIP Cilium logging guidelines:

https://docs.google.com/document/d/142qQ81YQbYysJ_GPQJtVRRnZ-zFCKH_2NsQ-FEnlK7Y/edit?usp=sharing

At a glance, if this is not likely to cause significant disruption to the overall operations of Cilium, then 'Warn' seems more appropriate.

Separately, I am mildly concerned that printing a warning or even info message with just "arping failed" and no implications to the user what the impact is likely to be or what they should do about it doesn't seem very helpful. I don't know what I'd do if I saw Cilium report "arping failed", let alone users with less familiarity about how Cilium operates.

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -997,8 +998,10 @@ func initializeFlags() {
+	flags.Bool(option.EnableBPFBypassFIBLookup, false, "Enable FIB lookup bypass optimization for nodeport reverse NAT handling")
 	option.BindEnv(option.EnableBPFBypassFIBLookup)
+	flags.MarkHidden(option.EnableBPFBypassFIBLookup)
+	flags.MarkDeprecated(option.EnableBPFBypassFIBLookup, fmt.Sprintf("This option will be removed in v1.12."))
```
We need to add a note in the upgrade guide. And I think we should clarify in the guide that it doesn't have any effect as of now (we usually keep the flag active until we actually remove it).

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1245,16 +1245,48 @@ working, take a look at `this KEP
+When kube-proxy replacement is enabled, Cilium does L2 neighbor discovery of nodes
+in the cluster. This is required for the service load-balancing in order to populate
+L2 addresses for backends for the fast-path where it is not possible to dynamically
+resolve neighbors on demand.
```
```suggestion
in the cluster. This is required for the service load-balancing to populate L2
addresses for backends since it is not possible to dynamically resolve neighbors
on demand in the fast-path.
```

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1245,16 +1245,48 @@ working, take a look at `this KEP
+entries. In some rare cases, Cilium 1.10 or earlier might have left stale entries behind
+in the neighbor table causing packets between some nodes to be dropped. To prevent from
+performing the neighbor discovery and instead relying on the Linux kernel to discover
+hosts on the same L2 network you can pass the ``--enable-l2-neigh-discovery=false``
```
```suggestion
in the neighbor table causing packets between some nodes to be dropped. To skip the
neighbor discovery and instead rely on the Linux kernel to discover neighbors, you can
pass the ``--enable-l2-neigh-discovery=false``
```

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1245,16 +1245,48 @@ working, take a look at `this KEP
+performing the neighbor discovery and instead relying on the Linux kernel to discover
+hosts on the same L2 network you can pass the ``--enable-l2-neigh-discovery=false``
+flag to the cilium-agent. However, note that relying on the Linux Kernel might also
+cause some packets to be dropped, for example, a NodePort request can be dropped on
```
You can't join two independent clauses with a just comma:
```suggestion
cause some packets to be dropped. For example, a NodePort request can be dropped on
```
(Or use a semi-colon.)

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1245,16 +1245,48 @@ working, take a look at `this KEP
+could happen if there is no L2 neighbor entry in the kernel (due to the entry being
+garbage collected or given that the neighbor resolution has not been done by the kernel).
+This is because it is not possible to drive the neighbor resolution from BPF programs
+in the fast-path e.g. at the XDP layer.
```
Overall, this new documentation section feels like exactly what we'd put in the release blog post. It's a bit unusual I think to have the documentation for a given version (v1.11) describe the behavior of previous versions.

borkmann replied:
I think it may be useful given most folks are still on v1.10 or earlier and these are two different approaches, so might be good to know what changed in terms of neighbor discovery. We can remove the before/after in 1.12, but I think it'd be good to have this info here.

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1245,16 +1245,48 @@ working, take a look at `this KEP
+and transparently use. In this case, the agent pushes down L3 addresses of new nodes
+joining the cluster as externally learned "managed" neighbor entries. For introspection,
+iproute2 displays them as "managed extern_learn". The "extern_learn" attribute is present
+such that the entries are not garbage collected by the kernel's neighboring subsystem
```
```suggestion
iproute2 displays them as "managed extern_learn". The "extern_learn" attribute prevents
garbage collection of the entries by the kernel's neighboring subsystem. Such "managed"
```

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1245,16 +1245,48 @@ working, take a look at `this KEP
+such that the entries are not garbage collected by the kernel's neighboring subsystem
+garbage collector. Such "managed" neighbor entries are dynamically resolved and periodically
+refreshed by the Linux kernel itself in case there is no active traffic for a certain period
+of time, meaning, the kernel attempts to always keep them in REACHABLE state. For Linux
```
Ditto, two independent clauses joined by a comma:
```suggestion
of time. That is, the kernel attempts to always keep them in REACHABLE state. For Linux
```

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1245,16 +1245,48 @@ working, take a look at `this KEP
+of time, then a Cilium agent controller triggers the Linux kernel-based re-resolution for
+attempting to keep them in REACHABLE state. The refresh interval can be changed if needed
+through a ``--arping-refresh-period=30s`` flag passed to the cilium-agent. The default
+period by the agent is ``30s`` which corresponds to the kernel's base reachable time.
```
```suggestion
period is ``30s`` which corresponds to the kernel's base reachable time.
```

pchaigno reviewed (approved):
Only minor comments from me. Logic looks good!

borkmann committed:
cilium, vendor: Pull in latest changes from vishvananda/netlink library

For the neighboring subsystem, we need the NTF_EXT_MANAGED flag for newer
kernels. Additionally, the netlink library is broken with regards to pushing
nil lladdresses into the kernel.

We addressed both recently in vishvananda/netlink.

While we were waiting for those two patches to land upstream, we temporarily
used cilium/netlink with included patches. The following was used for importing
cilium/netlink instead:

  go mod edit -replace github.com/vishvananda/netlink=github.com/cilium/netlink@master
  go mod tidy
  go mod vendor
  git add vendor/ go.mod go.sum && git commit -sa

However, just few days ago our PR was merged, so just regular update is already
sufficient (above instructions are kept in log for future reference). We used
the following for this patch:

  go get github.com/vishvananda/netlink@master
  go mod tidy
  go mod vendor
  git add vendor/ go.mod go.sum && git commit -sa

Related: https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=7482e3841d520a368426ac196720601687e2dc47
Related: https://github.com/vishvananda/netlink/pull/707
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
